File: Leetcode Questions/121-Best-Time-To-Buy-And-Sell-Stock/121-best-time-to-buy-and-sell-stock.cpp
#include<iostream>
#include<vector>
using namespace std;

int maxProfit(vector<int>& prices) {
    int max_profit = 0, min_price = INT16_MAX; 
    for(int x: prices) {
        min_price = min(x, min_price);
        max_profit = max(x - min_price, max_profit);
    }
    return max_profit;
}

int main() {
    vector<int> prices{7,1,5,3,6,4};
    cout<<maxProfit(prices);
    return 0; 
}

File: Leetcode Questions/121-Best-Time-To-Buy-And-Sell-Stock/121-best-time-to-buy-and-sell-stock.py
'''
You are given an array prices where prices[i] is the price of a given stock on the ith day.
You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
'''

def maxProfit(prices: list[int]) -> int:
    min_price = float('inf')
    max_profit = 0
    for x in prices:
        min_price = min(x, min_price)
        max_profit = max(x - min_price, max_profit)
    return max_profit

prices = [7,1,5,3,6,4]
print(maxProfit(prices))